### PR TITLE
TS: separate Window type for application under test

### LINKED ISSUE #7856: Correct Window type.
### Current behavior:

It's from [#6624](https://github.com/cypress-io/cypress/pull/6624#issue-382914365).

Previously this would type error:

```js
cy.window().then(window => window.eval('1'));
```

(`window` should have type `Window & typeof globalThis`.)

You may want to fix other instances of `Chainable<Window>` as well.

--------

It's [from the comment in #6624](https://github.com/cypress-io/cypress/pull/6624#issuecomment-595782820)

@OliverJAsh Regarding the issue we talked about — here's a suggestion for how we can _really_ get the right `Window` yielded by `cy.window()`.

If my suggestion falls outside the scope of this pull request, please let me know — I can open a separate issue or a pull request instead.

**Problem**

When you run a test with Cypress, there are two `Window` instances in play. 

1. The top `Window` object in which the Cypress UI lives, and
1. The `Window` that represents the tested application.

These are different `Window`s, but they share the same definition. If you define additional properties by using [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html), they will appear on both, which in most cases is incorrect and leads to bugs.

**Solution**

Allow users to define global variables separately.

In `cli/types/index.d.ts`, define an empty interface called `ApplicationWindow`.

```ts
interface ApplicationWindow {}
```

Change the definition for `cy.window` so that the yielded object includes the properties defined by `ApplicationWindow`.

```ts
window(options?: Partial<Loggable & Timeoutable>): Chainable<Window & typeof globalThis & ApplicationWindow>
```

**Usage**

The users can now define properties that exist on the yielded `window`, but not on the `window` accessed directly inside a Cypress test.

```ts
declare global {
  namespace Cypress {
    interface ApplicationWindow {
      __FOO__?: string;
    }
  }
}
```

image

_Originally posted by @karol-majewski in https://github.com/cypress-io/cypress/pull/6624#issuecomment-595782820_

### Desired behavior:

* Cypress `Window` definition should include `globalThis` definitions. 
* The type of Cypress `ApplicationWindow` can be extended. 

### Versions

Cypress 4.9.0.

### PR DESCRIPTION
**Pull request by sainthkh**
- Closes #7856

### User facing changelog

* Correct type for `Window` object
* Provide way to extend `Window` object type in Application Window. 

#### Breaking change

- The required minimum TypeScript version has been risen to 3.4.

#### Why was this change necessary?

TypeScript users cannot use valid typescript types. 

#### What is affected by this change?

N/A

#### Any implementation details to explain?

N/A

### How has the user experience changed?

N/A

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/2945
- [x] Has a PR for changes to kitchensink been opened? https://github.com/cypress-io/cypress-example-recipes/pull/515
- [x] Have API changes been updated in the [`type definitions`](../cli/types/cypress.d.ts)?
- [N/A] Have new configuration options been added to the [`cypress.schema.json`](../cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by sainthkh** [comment]
Flaky failures.

File: cli/types/cypress.d.ts [hunk 1]
```diff
@@ -107,6 +107,11 @@ declare namespace Cypress {
     fromAutWindow: WindowPosition & { x: number, y: number }
   }
 
+  /**
+   * The interface for user-defined properties in Window object under test.
+   */
+  interface ApplicationWindow {} // tslint:disable-line
```
Comment (bahmutov) [inline cli/types/cypress.d.ts:113]: can tslint disable specific error? Like in this case it is probably an empty interface error, right?

File: cli/types/cypress.d.ts [hunk 2]
```diff
@@ -2001,7 +2006,7 @@ declare namespace Cypress {
     })
     ```
      */
-    window(options?: Partial<Loggable & Timeoutable>): Chainable<Window>
+    window(options?: Partial<Loggable & Timeoutable>): Chainable<Window & typeof globalThis & ApplicationWindow>
```
Comment (bahmutov) [inline cli/types/cypress.d.ts:2009]: why add `globalThis`?

File: cli/types/index.d.ts [hunk 3]
```diff
@@ -1,33 +1,33 @@
...
-/// <reference path="./cypress-global-vars.d.ts" />
-/// <reference path="./cypress-type-helpers.d.ts" />
-/// <reference path="./cypress-expect.d.ts" />
+// Project: https://www.cypress.io
+// GitHub:  https://github.com/cypress-io/cypress
+// Definitions by: Gert Hengeveld <https://github.com/ghengeveld>
+//                 Mike Woudenberg <https://github.com/mikewoudenberg>
+//                 Robbert van Markus <https://github.com/rvanmarkus>
+//                 Nicholas Boll <https://github.com/nicholasboll>
+// TypeScript Version: 3.4
```
Comment (bahmutov) [inline cli/types/index.d.ts:7]: why did the entire file change?

Comment (bahmutov) [inline cli/types/index.d.ts:7]: can a PR to cypress-documentation for page https://on.cypress.io/typescript show the `ApplicationWindow` example AND minimum TypeScript version needed (3.4) in this case?

File: cli/types/cypress.d.ts (hunk 2 above)
Comment (OliverJAsh) [reply cli/types/cypress.d.ts:2009]: See https://github.com/cypress-io/cypress/pull/6624

Comment (OliverJAsh) [inline cli/types/cypress.d.ts:2009]: Are there other instances of `Chainable<Window>` that should also be updated?

Comment (OliverJAsh) [inline cli/types/cypress.d.ts:2009]: This also works:

```suggestion
    window(options?: Partial<Loggable & Timeoutable>): Chainable<typeof window & ApplicationWindow>
```

File: cli/types/cypress.d.ts (hunk 1 above)
Comment (sainthkh) [reply cli/types/cypress.d.ts:113]: I don't remember the message, but tslint doesn't allow empty interface(`{}`). 

Unfortunately, tslint doesn't have a way to disable only a rule like `/* eslint-disable a-rule */`.

File: cli/types/index.d.ts (hunk 3 above)
Comment (sainthkh) [reply cli/types/index.d.ts:7]: Interesting. I'll check if there's a change like line-ending.

**Comment by brian-mann** [comment]
Can we open an issue documenting this fix and why it's necessary. Also per @bahmutov's comments, it seems like there's a bit more work to do on this PR yeah?

**Comment by OliverJAsh** [comment]
I had a clear description in my original PR, but sadly that was closed: https://github.com/cypress-io/cypress/pull/6624

**Comment by sainthkh** [comment]
@OliverJAsh 

It's recommended to create an issue before submitting a PR in [the contributing guide](https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#pull-requests).

> If there is not an associated open issue, create an issue using our Issue Template.

Sometimes, ignoring guides can cause unwanted results. (And I also know that it's really tedious to read every contributing guide before contributing.)

Anyway, I opened an issue at #7856.

File: cli/types/cypress.d.ts (hunk 2 above)
Comment (sainthkh) [reply cli/types/cypress.d.ts:2009]: I copied and opened the issue at #7856.

File: cli/types/index.d.ts (hunk 3 above)
Comment (sainthkh) [reply cli/types/index.d.ts:7]: Line ending was changed at d719f6618e2f893afb66925b4e4611576442591d. Fixed.

**Comment by sainthkh** [comment]
@bahmutov @brian-mann 

There are 3 more things to do:

- [ ] Create `AUTWindow` type and replace the appropriate `Window` types in `cli/types/cypress.d.ts` with it. 
- [ ] Create a new recipe at cypress-example-recipes to showcase out-of-the-box TypeScript, how to check typescript types and ApplicationWindow type extension
- [ ] Create a new PR at https://on.cypress.io/typescript cypress-doc that the required minimum TypeScript version has been risen to 3.4

**Comment by brian-mann** [comment]
@sainthkh proceed ahead - we are trying to release `5.0.0` in the next day or two and want to get this in.

**Comment by jennifer-shehane** [comment]
@sainthkh `cy.window()` is not the only command to return the `Window`, will the other command's definitions be extended? The other commands:

- `cy.go()`
- `cy.reload()`
- `cy.visit()`

**Comment by jennifer-shehane** [comment]
I've opened a PR for the docs to update the min required version to 3.4+ https://github.com/cypress-io/cypress-documentation/pull/2945

**Comment by sainthkh** [comment]
It seems that kitchensink needs update after #7782. [(link)](https://app.circleci.com/pipelines/github/cypress-io/cypress/12124/workflows/add4d714-a4fa-4896-91e8-0d64ba96f663/jobs/383685/steps)

**Pull request merged**